components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/model/auth/service/AuthServiceRequestWrapper.java

```diff
@@ -0,0 +1,150 @@
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatorData;
+import org.wso2.carbon.identity.application.authentication.framework.util.FrameworkConstants;
+import org.wso2.carbon.identity.application.authentication.framework.util.auth.service.AuthServiceConstants;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+
+/**
+ * A HttpServletRequest wrapper to be used in authentication service.
+ */
+public class AuthServiceRequestWrapper extends HttpServletRequestWrapper {
+
+    private Map<String, String[]> parameters = new HashMap<>();
+    private List<AuthenticatorData> authInitiationData = new ArrayList<>();
+
+    public AuthServiceRequestWrapper(HttpServletRequest request, Map<String, String[]> parameters) {
+
+        super(request);
+        this.parameters = parameters;
+        setSessionDataKey(parameters);
+        request.setAttribute(AuthServiceConstants.AUTH_SERVICE_AUTH_INITIATION_DATA, authInitiationData);
+        skipNonceCookieValidation();
+    }
+
+    @Override
+    public String getParameter(String name) {
+
+        if (this.parameters.containsKey(name) && this.parameters.get(name).length > 0) {
+            return this.parameters.get(name)[0];
+        }
+        return super.getParameter(name);
+    }
+
+    @Override
+    public Map<String, String[]> getParameterMap() {
+
+        Map<String, String[]> paramMap = new HashMap<>(this.parameters);
+        Map<String, String[]> paramsFromSuper = super.getParameterMap();
+        paramsFromSuper.forEach((key, value) -> {
+            if (!paramMap.containsKey(key)) {
+                paramMap.put(key, value);
+            }
+        });
+
+        return paramMap;
+    }
+
+    @Override
+    public Enumeration<String> getParameterNames() {
+
+        ArrayList<String> paramNames = new ArrayList<>(this.parameters.keySet());
+        Enumeration<String> paramsFromSuper = super.getParameterNames();
+
+        while (paramsFromSuper.hasMoreElements()) {
+            String paramName = paramsFromSuper.nextElement();
+            if (!paramNames.contains(paramName)) {
+                paramNames.add(paramName);
+            }
+        }
+
+        return Collections.enumeration(paramNames);
+    }
+
+    @Override
+    public String[] getParameterValues(String name) {
+
+        if (this.parameters.containsKey(name)) {
+            return this.parameters.get(name);
+        }
+
+        return super.getParameterValues(name);
+    }
+
+    /**
+     * Get the authentication initiation data related to
+     * any authenticators engaged in the request flow.
+     *
+     * @return List of {@link AuthenticatorData} objects.
+     */
+    public List<AuthenticatorData> getAuthInitiationData() {
```

Comment:
authInitiationData looks confusing to me, gives the feeling it contains, the data related to authentication flow initiation from the initial request